File: 0150-evaluate-reverse-polish-notation/solution.py
class Solution(object):
    def evalRPN(self, tokens):
        """
        :type tokens: List[str]
        :rtype: int
        """
        
        S = collections.deque()
        
        for i in xrange(len(tokens)):
            if tokens[i] == "+":
                a = int(S.popleft())
                b = int(S.popleft())
                S.appendleft(a+b)
            elif tokens[i] == "-":
                a = int(S.popleft())
                b = int(S.popleft())
                S.appendleft(b-a)
            elif tokens[i] == "*":
                a = int(S.popleft())
                b = int(S.popleft())
                S.appendleft(b*a)
            elif tokens[i] == "/":
                a = float(S.popleft())
                b = float(S.popleft())
                S.appendleft(int(b/a))
            else:
                S.appendleft(tokens[i])
        
        return int(S.popleft())

File: 0150-evaluate-reverse-polish-notation/solution.cpp
class Solution {
public:
    int evalRPN(vector<string>& tokens) {
        stack<int> S;

        for (int i = 0; i < tokens.size(); ++i) {
            if (tokens[i] == "+" || tokens[i] == "-" || tokens[i] == "*" || tokens[i] == "/") {
                int b, a;
                b = S.top(); S.pop();
                a = S.top(); S.pop();

                
                if ( tokens[i] == "+") { 
                    S.push(a + b);
                } else if (tokens[i] ==  "-") {
                    S.push(a - b);
                } else if (tokens[i] == "*") {
                    S.push(a*b);
                } else {
                    S.push(a/b);
                }
            } else {
                S.push(stoi(tokens[i]));
            }
        }

        return S.top();
    }
};
